### PR TITLE
Ask for confirmation, before loading burn after reading pastes

### LINKED ISSUE #1039: German message overlapping
please see attached the formatting for "The recipient could get to know your timezone, would you like to convert the time to UTC?

grafik

### LINKED ISSUE #1191: Changing language invalidates URL of existing paste in browser
- [x] Yes, I have read [the FAQ](https://github.com/PrivateBin/PrivateBin/wiki/FAQ) and I found no solution/answer there.

## Steps to reproduce

1. Create a new paste; "English" is selected in the combo at top right of the page
2. Press "Send", get a link: https://privatebin.net/?2a7bea9817c47814#A3WdsTpYXocSAKMdFahpsDyKMmGvzQWUw5tEDaFhe2Mu
3. Re-start browser, enter link above; everything works fine
4. Change language to another one, e.g. "Deutsch" (or, just press "English" again)

### What happens

The URL in the address bar gets shortened to https://privatebin.net/?2a7bea9817c47814# (i.e. the part after `#` is skipped), and a message is displayed:

Cannot decrypt paste: Decryption key missing in URL (Did you use a redirector or an URL shortener which strips part of the URL?)

### What should happen

I am not sure whether this behaviour is intended; but imho it should be possible to read an existing paste eben if one changes the language of the ui.

## Basic information

**Server address**:

https://privatebin.net/

**Browser:**
Firefox 119.0, Edge 118.0.2088.76

**PrivateBin version:** 1.6.0

### PR DESCRIPTION
**Pull request by elrido**
This PR is derived from #1231 but with a slightly less complex goal.

When creating burn after reading pastes, a "?" is inserted at the start of the URL fragment (.../?aaaa#bbbb becomes .../?aaaa#**?**bbbb). This way the server doesn't get sent this, though it still knows that the paste is of the burn after reading type. If this character is detected upon loading a paste, the user is asked to confirm loading it, by clicking a button:

load confirmation dialog

While this is done automatically for burn after reading pastes, it doesn't prevent anyone from manually adding it to a regular URL to get it to ask for non burn after reading pastes (hence the slightly vague message) or removing it from burn after reading URL to instantly load it, if this isn't required.

## Changes
* Ask before loading burn after reading pastes
* fixes #1039 - email buttons overlapping in some languages, by moving the third button up to a close button (X): 
email dialog
* fixes #1191 - language change URL mangling by preventing the default click action
* adds focus to password input in modal
* prevents needless reload on visiting default URL
* moving the close button in the QR code dialog up to a close button (X):
qr code dialog

## ToDo
* [ ] discuss if we could remove the isBadBot() mechanism from legacy.js
* [ ] discuss the wording of the new dialog - I feel it sounds a bit off, but had no better ideas.

### CODE REVIEW

File: js/privatebin.js [hunk 1]
```diff
@@ -77,6 +77,13 @@ jQuery.PrivateBin = (function($, RawDeflate) {
         }
     };
 
+    /**
+     * URL fragment prefix requiring load confirmation
+     *
+     * @private
+     */
+    const loadConfirmPrefix = '#?';
```
Comment (rugk) [inline js/privatebin.js:85]: Given the ambiguity such a character is used in URLs for encoding query params etc. I am not sure whether this is the best character to find after the anchor/fragement part? https://en.wikipedia.org/wiki/Percent-encoding
`~` seems to be a safe character?

[But apparently spec allows any ASCII chars](https://url.spec.whatwg.org/#concept-url-fragment)? Just saying this, because nowadays [there can be crazy stuff in there](https://developer.mozilla.org/en-US/docs/Web/Text_fragments).

I have nothing against this, we should just think of potential problems of this schema etc. now, before all URLs are in use… :upside_down_face: 

Ah or did you use that because of the meaning as question mark?

File: tpl/bootstrap.php [hunk 2]
```diff
@@ -123,46 +123,52 @@
...
+					<div class="modal-body text-center">
+						<button id="loadconfirm-open-now" type="button" class="btn btn-success" data-dismiss="modal"><span class="glyphicon glyphicon-download"></span> <?php echo I18n::_('Yes, load it') ?></button>
+					</div>
+				</div>
+			</div>
+		</div>
 <?php
 if ($QRCODE) :
 ?>
-		<div id="qrcodemodal" tabindex="-1" class="modal fade" aria-labelledby="qrcodemodalTitle" role="dialog" aria-hidden="true">
```
Comment (rugk) [inline tpl/bootstrap.php:129]: Seems you removed `aria-labelledby`. Was this intentional? And does not this negatively affect accessibility?

**Review by rugk — APPROVED** [review]
Generally LGTM and a good idea and simple enough solution!

**Comment by rugk** [comment]
> discuss if we could remove the isBadBot() mechanism from legacy.js

It's okay. The only thing to clarify could be to label the button explicitly "Yes, load and burn" or so.

> it doesn't prevent anyone from manually adding it to a regular URL to get it to ask for non burn after reading pastes (hence the slightly vague message)

That is a good idea and feature, actually, IMHO!
However, for a real proper localization, we can of course detect on the client-side if the paste was a self-destroy one (we have that information IIRC) and show a different message (aka more generic one), then.

File: js/privatebin.js (hunk 1 above)
Comment (elrido) [reply js/privatebin.js:85]: This is certainly worth thinking over twice. Yes, I picked the question mark simply because the effect will be to ask the user, so adding it triggers a question. My only other concern was to not use any hexadecimal characters (0-9, a-f or A-F) and to keep it to just one character since a lot of folks don't like long URLs. But I think we could really use any other ASCII character, also upper or lower case letters, punctuation, etc. I would like it to remain a visible character and something not easily lost when copy pasting across constrained media or communication channels. That would be an argument against the question mark, as some parsers may not allow it or think it is where they can append URL arguments after (thinking of certain social media platforms adding tracking parameters). So here are some more ideas:
- example.com/?aaaa**#q**bbbb (as in "question")
- example.com/?aaaa**#Q**bbbb
- example.com/?aaaa**#-**bbbb (small)
- example.com/?aaaa**#>**bbbb
- example.com/?aaaa**#G**bbbb (first non-hex letter)
- example.com/?aaaa**#i**bbbb (least visible, might get us in trouble with a certain fruit company)

Certainly worth pondering some more.

File: tpl/bootstrap.php (hunk 2 above)
Comment (elrido) [reply tpl/bootstrap.php:129]: It was - it only makes sense to have that if it does reference an ID that contains an area label. Those IDs do not exist, so it didn't actually add any value to visually impaired readers. I did add actual area labels to the new "Close" buttons and a proper title to the QR code modal. Before this change, the modal would open and only read "Close" for the close button, but the inserted QR code doesn't provide any readable content. Now at least it will read "QR code - pause - Close".

**Comment by elrido** [comment]
> However, for a real proper localization, we can of course detect on the client-side if the paste was a self-destroy one (we have that information IIRC) and show a different message (aka more generic one), then.

As a recipient of the link, you only learned if the paste is burn after reading after loading it. The server knows (so they can delete it), but we would need to add an additional non-destructive API for the client to learn this, and that of course undermines the concept of not leaving a trace (but would let us add a counter: "This paste has been visited but not opened X amount of times") - Anyhow, I don't think it would be worth the extra risk and complexity.

File: js/privatebin.js (hunk 1 above)
Comment (elrido) [reply js/privatebin.js:85]: I realize my above statement reg. avoiding hexadecimal characters is not correct: In the URL fragment, we use base58 (v2 pastes) or base64 (v1 pastes) encoding. This means we should avoid using a-z, A-Z, 0-9, +, / and = (for base64). Base58 uses a smaller alphabet, but someone could modify an older v1 paste URL to ask the question, so we better avoid all of these. This would restrict us to punctuation like !, dash, underscore, semicolons and colons, brackets, ...

Well, or we go unicode and use an emoji ;-) It would make the URL longer and there are some risks when shared over services that would apply URL encoding to it. So probably best to stick to ASCII.

Comment (rugk) [reply js/privatebin.js:85]: For me a simple `-` or `_` or so would be good IMHO. < and > could cause problems with (Markdown) URL parsers.

Comment (elrido) [reply js/privatebin.js]: I've now updated the prefix to a dash (`-`) and also re-fixed the password modal focusing.

Found that it would still loose focus shortly after getting set. Since the debugger pointed into jquery code, not ours, I applied the workaround with a timeout to focus back on it. Happens in Firefox as well as Chrome. :-/ Hope we can drop this when moving on to a newer bootstrap version.

**Review by rugk — APPROVED** [review]

**Comment by rugk** [comment]
> As a recipient of the link, you only learned if the paste is burn after reading after loading it

Ah okay, thanks. Then I mis-remembered that this may have been saved in the authenticated data part of the encryption.

**Comment by nain-F49FF806** [comment]
Hi, the top comment says a character "?" is inserted after the "#", but the [code seems to use "#-" as prefix](https://github.com/PrivateBin/PrivateBin/pull/1237/files#diff-4c30f267da1a33c8d947890833316f48f2c58b89bc0ae567a469056ce98b96e5R80-R86). 

Could you offer clarification on which character third party clients should look for?
Also, perhaps we should document the full structure of the fragment somewhere on the wiki.

**Comment by elrido** [comment]
> Hi, the top comment says a character "?" is inserted after the "#", but the [code seems to use "#-" as prefix](https://github.com/PrivateBin/PrivateBin/pull/1237/files#diff-4c30f267da1a33c8d947890833316f48f2c58b89bc0ae567a469056ce98b96e5R80-R86).
> 
> Could you offer clarification on which character third party clients should look for? Also, perhaps we should document the full structure of the fragment somewhere on the wiki.

That was the initial suggestion, but after a [discussion and considering various options and implication in the code review](https://github.com/PrivateBin/PrivateBin/pull/1237#discussion_r1480438726), we ended up changing it to the less ambiguous dash.

**Comment by nain-F49FF806** [comment]
Thanks for the clarification! And the URL format description is very helpful. :)

I understand this is already in production. But since it's relatively new, I wanted to ask what you think of having the character as a suffix, rather than prefix? 

Were there any strong arguments for having the character before the key and not after?

If it's in the suffix, the scheme could be extended for other custom options. Clients could always rely on the key being present first after the `#`, and to ignore anything after the `-` character if present.

**Comment by nain-F49FF806** [comment]
Something like :

Without custom options (current default):

```
paste_url = privatebin_url + paste_id + "#" + encoded_key
```

With custom [client-side] options:
```
paste_url = privatebin_url + paste_id + "#" + encoded_key + client_options 
```

Where client options can be defined separately, and supporting them would be optional.

```
client_options:
    '-' : warn before read
```

**Comment by elrido** [comment]
I don't see how a character at the start or end makes it any simpler or more difficult to parse. Finding the optional extra char at the beginning or the end just changes your regex pattern or if you check first for the char and then for the base58 or the other way around.

Hence I would not want to change it now, as it is already being used in the wild. Changing it would mean that your client now has to support both the 1.7 version of the URL and the new version you propose.

**Comment by nain-F49FF806** [comment]
> I don't see how a character at the start or end makes it any simpler or more difficult to parse. 

I agree, for a *single character* it doesn't make a difference. What I was trying to convey was, if in the future there are *more* option characters, then with prefix, clients would have to update to support each new character, or they will break. 

Am sorry for not being able to convey properly before, let me try again. 

Say Privatebin 1.9 wants to introduce a character to warn that paste contains large attachment (`🐘`), again as prefix.
Now clients would have to dynamically search for presence of either this new `🐘` chararater or the `-` character or both, before getting to the key.  *If they don't know yet* about the `🐘` character, then parsing for base58_key may break unless we do some complicated regex matching on base58 characters. 

With suffix system, you could simply define fragment linearly with anchor char (`#`) as origin. 

| location relative to `#`/`#-` | what it is                           | decode                 | introduced in   |   |
|--------------------------|--------------------------------------|------------------------|-----------------|---|
| char 1 - 45              | base58_key                           | base58 decode          | -               |   |
| char 46                  | warn about burn on read              | `-` = on, `_` = off    | Privatebin v1.7.2 |   |
| char 47                  | warn that paste has large attachment | `-`  = on,  `_`  = off | Privatebin v1.9 |   |
| char 48                  | ?                                    | ?                      | ?               |   |

or treating fragment as csv 

| location relative to `#`/`#-` | what it is                           | decode                 | introduced in   |   |
|--------------------------|--------------------------------------|------------------------|-----------------|---|
| item 1               | base58_key                           | base58 decode          | -               |   |
| item 2                 | warn about burn on read              | `-` = on  | Privatebin v1.7.2 |   |
| item 3                  | warn that paste has large attachment | `-`  = on| Privatebin v1.9 |   |
| item 4                  | ?                                    | ?                      | ?               |   |

clients can then support upto index they know. 
And not break on future append only updates to fragment structure.

> Q: Isn't this also possible to do by reverse indexing, by counting from end of url ?

Having to index from behind is brittle since other programs or browsers can add spurious chars at the end of url.
So indexing forward from the `#` would be more robust.

Basically, my suggestion is to keep the base58_key and any other option chars at fixed positive positions relative to the `#` /`#-` prefix in future updates

**Comment by elrido** [comment]
I would not recommend to index from behind either, because, as we had to learn, lot's of tracking-tools in chat-apps or "social" media will automatically add extra "GET-parameters" (ampersand followed by some key=value pairs) to the end of any URL-like string, even if that ends up being appended to the fragment and not to the query part of the URL.

Instead I would parse it as follows:
1. is the first character of the fragment a dash?
2. If yes, ignore it and shift one character forward (or take note of this and do what you want with that information, any use of this is client context specific - it doesn't guarantee you that it is a burn-after-reading paste, for example)
3. from where you now are, can the next 42 characters or the characters till the end of the fragment or till the next ampersand (whichever comes first) decode as base58 into 32 bytes ([zero-padded at the front, if there are less than 32 bytes](https://github.com/PrivateBin/PrivateBin/blob/ec02afca0469017a77fa5c8b952241c879883882/js/privatebin.js#L1543-L1545))?
4. if yes, that is the paste key, else you do not have a valid privatebin URL

Granted, I may have gotten exposed to too much C and too many char arrays, in recent years.

I'm not aware of any plans to introduce any other special characters at this time and have so far myself not considered any other such items that we must pass via URL and can't introduce as paste parameters, either in the encrypted payload or the clear text adata.

**Comment by rugk** [comment]
One argument that would indeed be a good one (IMHO) for putting the character at the end is that a knowledgable user can easier modify it (given some weird things do not add stuff there, but this is more a rare problem IMHO).

* if symbol is at the end: Press <kbd>Ctrl</kbd>+<kbd>L</kbd>, then <kbd>→</kbd> and then <kbd>Backspace</kbd>. <kbd>Enter</kbd> to access the new URL or copy it.
* if symbol is at the start: Press <kbd>Ctrl</kbd>+<kbd>L</kbd>, then <kbd>←</kbd>, then <kbd>Ctrl</kbd>+<kbd>→</kbd> 4 times and then <kbd>Backspace</kbd>. <kbd>Enter</kbd> to access the new URL or copy it.

Similarly you can add the `-` manually, if you want.

Also, it _may_ be easier to see at the end instead of at the beginning – which in the URL is "inside some position in some wild letters".
So if we can still change it, I would also argue place it at the end may be a nicer idea.

Also that behavior is known e.g. from URL shorteners, where https://is.gd/vrlkOW- gives you a preview page and shows the URL first, while https://is.gd/vrlkOW redirects directly.

**Comment by elrido** [comment]
This was already merged and has been in release 1.7.0 and 1.7.1 - if we change it, third party clients will have to support three styles, for the classic URL, the 1.7.x URL and the future URL.

**Comment by nain-F49FF806** [comment]
Update to this discussion. 

For future versions, we could support such client-side features using [`fragmet directives`](https://wicg.github.io/scroll-to-text-fragment/#the-fragment-directive).
This could be the standard way to add support for such features, without breaking backwards compatibility.

**For example burn on read could be indicated as**:

`#9rsiGhAeDjdCScSQQeqpnXHZsfgJohFbRoGfehkMKZZt:~:mayburn`
or simply

`#9rsiGhAeDjdCScSQQeqpnXHZsfgJohFbRoGfehkMKZZt:~:🔥`

**Example burn on read, and big attachment** :

`#9rsiGhAeDjdCScSQQeqpnXHZsfgJohFbRoGfehkMKZZt:~:mayburn&bigattach`
or simply
`#9rsiGhAeDjdCScSQQeqpnXHZsfgJohFbRoGfehkMKZZt:~:🔥&🐘`

**Backward compatibility**: 
The optional `-` at the start of fragment (before key) **should still be supported as special case**, however.

`#-9rsiGhAeDjdCScSQQeqpnXHZsfgJohFbRoGfehkMKZZt`

`#-9rsiGhAeDjdCScSQQeqpnXHZsfgJohFbRoGfehkMKZZt:~:🐘`

**Comment by nain-F49FF806** [comment]
**Note**: We may have to change the delimeter (`:~:`) slightly.  As iiuc browsers may strip it off (along with the directives) after their processing and not send it to page scripts.
